sycl/include/sycl/ext/intel/esimd/detail/memory_intrin.hpp

```diff
@@ -500,17 +518,59 @@ __esimd_svm_atomic1(__ESIMD_DNS::vector_type_t<uint64_t, N> addrs,
 {
   __ESIMD_DNS::vector_type_t<Ty, N> retv;
 
-  for (int i = 0; i < N; i++) {
-    if (pred[i]) {
-      Ty *p = reinterpret_cast<Ty *>(addrs[i]);
+  for (int AddrIdx = 0; AddrIdx < N; AddrIdx++) {
+    if (pred[AddrIdx] == 0) {
+      // Skip Output vector elements correpsonding to
+      // predicates whose value is zero
+      continue;
+    }
 
-      switch (Op) {
-      case __ESIMD_NS::atomic_op::add:
-        retv[i] = __ESIMD_DNS::atomic_add_fetch<Ty>(p, src0[i]);
-        break;
-      default:
-        __ESIMD_UNSUPPORTED_ON_HOST;
+    // Keeping original values for return
+    retv[AddrIdx] = *((Ty *)addrs[AddrIdx]);
+
+    if constexpr (Op == __ESIMD_NS::atomic_op::store) {
+      __ESIMD_DNS::atomic_store<Ty>((Ty *)addrs[AddrIdx], src0[AddrIdx]);
+    } else if constexpr (Op == __ESIMD_NS::atomic_op::add) {
+      __ESIMD_DNS::atomic_add_fetch<Ty>((Ty *)addrs[AddrIdx], src0[AddrIdx]);
+    } else if constexpr (Op == __ESIMD_NS::atomic_op::sub) {
+      __ESIMD_DNS::atomic_sub_fetch<Ty>((Ty *)addrs[AddrIdx], src0[AddrIdx]);
+    } else if constexpr (Op == __ESIMD_NS::atomic_op::minsint) {
+      __ESIMD_DNS::atomic_min<Ty>((Ty *)addrs[AddrIdx], src0[AddrIdx]);
+    } else if constexpr (Op == __ESIMD_NS::atomic_op::maxsint) {
+      __ESIMD_DNS::atomic_max<Ty>((Ty *)addrs[AddrIdx], src0[AddrIdx]);
+    } else if constexpr (Op == __ESIMD_NS::atomic_op::min) {
+      if constexpr (!__ESIMD_DNS::is_fp_type<Ty>::value) {
+        __ESIMD_DNS::atomic_min<Ty>((Ty *)addrs[AddrIdx], src0[AddrIdx]);
+      }
+    } else if constexpr (Op == __ESIMD_NS::atomic_op::max) {
+      if constexpr (!__ESIMD_DNS::is_fp_type<Ty>::value) {
+        __ESIMD_DNS::atomic_max<Ty>((Ty *)addrs[AddrIdx], src0[AddrIdx]);
+      }
+    } else if constexpr (Op == __ESIMD_NS::atomic_op::fadd) {
```

Comment:
fadd and integer add, fsub and integer sub, fmax and umax/smax, fmin and umin/smin cases can be merged

Reply:
Done.